opds_springer/feed_generator.py

```diff
@@ -1,6 +1,203 @@
+            },
+            "links": self.opds_response_links(page_number),
+            "publications": publications,
+        }
+
+    def opds_response_links(self, page_number):
+        """Formats links data for OPDS response.
+
+        Args:
+            page_number (int): current page number
+
+        Returns:
+            list: list of dictionaries containing link information
+
+        """
+
+        links = []
+        self_rel = "self"
+        first_rel = "first"
+        last_rel = "last"
+        prev_rel = "prev"
+        next_rel = "next"
+        if page_number == 1:
+            self_rel = ["self", "first"]
+            first_rel = None
+            prev_rel = None
+        if page_number == 2:
+            first_rel = ["prev", "first"]
+            prev_rel = None
+        if page_number + 1 == self.total_pages:
+            last_rel = ["next", "last"]
+            next_rel = None
+        if page_number == self.total_pages:
+            self_rel = ["self", "last"]
+            next_rel = None
+            last_rel = None
+        if self_rel:
+            links.append(self.opds_response_link(page_number, self_rel))
+        if first_rel:
+            links.append(self.opds_response_link(1, first_rel))
+        if last_rel:
+            links.append(self.opds_response_link(self.total_pages, last_rel))
+        if next_rel:
+            links.append(self.opds_response_link(page_number + 1, next_rel))
+        if prev_rel:
+            links.append(self.opds_response_link(page_number - 1, prev_rel))
+        return links
+
+    def opds_response_link(self, page_number, rel):
+        """Formats a link.
+
+        Args:
+            page_number (int): current page number
+            rel (str): e.g., 'last'
+
+        Returns:
+            dict: link information
+
+        """
+        return {
+            "rel": rel,
+            "href": f"https://ebooks-test.library.columbia.edu/static-feeds/springer/springer_test_feed_{page_number}.json",
+            "type": "application/opds+json",
+        }
+
+    def get_books_from_db(self):
+        """Gets all book records.
+
+        Yields:
+            obj: book record
+        """
+        result = session.execute(select(Book))
+        for book in result.all():
+            yield book[0]
+
+
+class BookOPDS(object):
+    def create_json(self, book):
+        """Creates dictionary of OPDS book data.
+
+        Args:
+            book (obj): book record
+
+        Returns:
+            dict: book data
+        """
+        self.book = book
+        book_dict = {
+            "metatadata": self.metadata(),
+            "images": self.images(),
+            "links": self.links(),
+        }
+        return book_dict
+
+    def metadata(self):
+        """Creates dictionary of book metadata."""
+        metadata = {
+            "identifier": f"https://dx.doi.org/{self.book.book_id}",
+            "modified": datetime.utcnow().isoformat(),
```

Comment:
I think this commits us to (on the Palace side) a full reload monthly, right? That might make sense, but I want to make sure we've been explicit about the process here.

Reply:
It actually makes sense as a matter of process, if we are treating the db like a cache, and the opds as a transformation of the kbart. Storing a modified date on the book entity would mean that the kbart data *also* has to be stored and sorted - which doesn't seem quite like the process we want.

Reply:
But then how is the cache populated? This seems like it will report for the feed size the number of books historically encountered, but then actually have the books in the kbart feed. Does this present a pagination problem?